### PR TITLE
MGMT-19498: Adding accelerators configuration file to the node-exporter

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
1) adding ConfigMap that contains the yaml structure describing
   monitored accelerators per vendor/model
2) mapping accelerators ConfigMap into the node-exporter pods 3) adding env variable that contains the exact path for the mapped
   ConfigMap

* [ ] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

### CODE REVIEW

File: assets/node-exporter/accelerators-configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        - pciID: 0x2235
+          pciName: A40
+        - pciID: 0x1df5
+          pciName: V100
+        - pciID: 0x20f1
+          pciName: A100 40G
+        - pciID: 0x1ff2
+          pciName: T400 4GB
+        - pciID: 0x1eb8
+          pciName: Tesla T4
```
Comment (romfreiman) [inline assets/node-exporter/accelerators-configmap.yaml:22]: why not to add here more ids?

Comment (yevgeny-shnaidman) [reply assets/node-exporter/accelerators-configmap.yaml:22]: i want the code change to be approved first. Once that is done, then i will add the rest of the accelerators

File: jsonnet/components/node-exporter.libsonnet [hunk 2]
```diff
@@ -3,6 +3,31 @@ local textfileVolumeName = 'node-exporter-textfile';
 local tlsVolumeName = 'node-exporter-tls';
 local wtmpPath = '/var/log/wtmp';
 local wtmpVolumeName = 'node-exporter-wtmp';
+local configDir = '/var/node_exporter/config';
+local configVolumeName = 'node-exporter-config';
+local acceleratorsConfigFileName = 'accelerators-config.yaml';
+local acceleratorsConfigMapName = 'accelerators-configmap';
+local acceleratorsConfigData = |||
```
Comment (simonpasquier) [inline jsonnet/components/node-exporter.libsonnet:10]: you can use `std.manifestYamlDoc()` to convert a valid jsonnet object to its YAML representation. It will avoid any indentation issue.
https://jsonnet.org/ref/stdlib.html

File: jsonnet/components/node-exporter.libsonnet [hunk 3]
```diff
@@ -290,6 +322,10 @@ function(params)
                           name: 'DBUS_SYSTEM_BUS_ADDRESS',
                           value: 'unix:path=/host/root/var/run/dbus/system_bus_socket',
                         },
+                        {
+                          name: 'ACCELERATORS_CONFIG_FILE',
```
Comment (simonpasquier) [inline jsonnet/components/node-exporter.libsonnet:326]: I don't think that we need to make it configurable given that it's a pure downstream feature: the accelerator collector config file location can be hardcoded in node_exporter and mounted by CMO at the expected path.

File: pkg/manifests/manifests.go [hunk 4]
```diff
@@ -138,6 +138,7 @@ var (
 	NodeExporterMinimalServiceMonitor      = "node-exporter/minimal-service-monitor.yaml"
 	NodeExporterPrometheusRule             = "node-exporter/prometheus-rule.yaml"
 	NodeExporterKubeRbacProxySecret        = "node-exporter/kube-rbac-proxy-secret.yaml"
+	NodeExporterConfigMap                  = "node-exporter/accelerators-configmap.yaml"
```
Comment (jan--f) [inline pkg/manifests/manifests.go:141]:
```suggestion
	NodeExporterAcceleratorsConfigMap                  = "node-exporter/accelerators-configmap.yaml"
```
Maybe a more descriptive name?

File: jsonnet/components/node-exporter.libsonnet (hunk 3 above)
Comment (yevgeny-shnaidman) [reply jsonnet/components/node-exporter.libsonnet:326]: Using the environment variable means that we determine the mounting path only at CMO, and there is no need to synchronize between CMO and node-exporter. IMHO in case it will be changed in the future for any reason, using ENV will allow us to avoid synchronization bugs

File: pkg/manifests/manifests.go (hunk 4 above)
Comment (yevgeny-shnaidman) [reply pkg/manifests/manifests.go:141]: Fixed

File: jsonnet/components/node-exporter.libsonnet (hunk 2 above)
Comment (yevgeny-shnaidman) [reply jsonnet/components/node-exporter.libsonnet]: Fixed

**Comment by yevgeny-shnaidman** [comment]
@jan--f @simonpasquier , fixed and answered the remarks, please take a look again

File: jsonnet/components/node-exporter.libsonnet (hunk 3 above)
Comment (machine424) [reply jsonnet/components/node-exporter.libsonnet:321]: I agree with Simon, hardcoding it would be better as we control the whole chain, and in case of a sync issue, we should be expecting the collector and the tests to yell.
(But if you think this is still needed I'd go for a cli arg for consistency, like `--collector.textfile.directory=/var/node_exporter/textfile` e.g.)

File: jsonnet/versions.yaml [hunk 5]
```diff
@@ -15,7 +15,7 @@ repos:
     thanos: openshift/thanos
 versions:
     alertmanager: 0.27.0
-    kubeRbacProxy: 0.18.1
+    kubeRbacProxy: 0.18.2
```
Comment (machine424) [inline jsonnet/versions.yaml:18]: could you revert this? as it's not related to the PR. (in case the PR is reverted we don't want this to get reverted e.g.)
We'll take case of it in another PR. `versions` job isn't required you can keep it failing.

File: jsonnet/components/node-exporter.libsonnet [hunk 6]
```diff
@@ -3,6 +3,26 @@ local textfileVolumeName = 'node-exporter-textfile';
 local tlsVolumeName = 'node-exporter-tls';
 local wtmpPath = '/var/log/wtmp';
 local wtmpVolumeName = 'node-exporter-wtmp';
+local configDir = '/var/node_exporter/config';
+local configVolumeName = 'node-exporter-config';
+local acceleratorsConfigFileName = 'accelerators-config.yaml';
+local acceleratorsConfigMapName = 'accelerators-configmap';
```
Comment (machine424) [inline jsonnet/components/node-exporter.libsonnet:9]: to reduce the risk of any collision and to provide as much context as possible I'd go with `node-exporter-accelerators-collector-config` (this is also consistent with other naming)

File: jsonnet/components/node-exporter.libsonnet [hunk 7]
```diff
@@ -324,6 +355,18 @@ function(params)
                   secretName: 'node-exporter-kube-rbac-proxy-config',
                 },
               },
+              {
+                name: configVolumeName,
+                configMap: {
+                  name: acceleratorsConfigMapName,
+                  items: [
+                    {
+                      key: 'accelerators-config.yaml',
```
Comment (machine424) [inline jsonnet/components/node-exporter.libsonnet:364]:
```suggestion
                      key: 'accelerators-collector-config.yaml',
```

File: pkg/manifests/manifests.go [hunk 8]
```diff
@@ -138,6 +138,7 @@ var (
 	NodeExporterMinimalServiceMonitor      = "node-exporter/minimal-service-monitor.yaml"
 	NodeExporterPrometheusRule             = "node-exporter/prometheus-rule.yaml"
 	NodeExporterKubeRbacProxySecret        = "node-exporter/kube-rbac-proxy-secret.yaml"
+	NodeExporterAcceleratorsConfigMap      = "node-exporter/accelerators-configmap.yaml"
```
Comment (machine424) [inline pkg/manifests/manifests.go:141]:
```suggestion
	NodeExporterAcceleratorsConfigMap      = "node-exporter/accelerators-collector-configmap.yaml"
```

File: pkg/tasks/nodeexporter.go [hunk 9]
```diff
@@ -94,6 +94,15 @@ func (t *NodeExporterTask) Run(ctx context.Context) error {
 		return fmt.Errorf("reconciling node-exporter Service failed: %w", err)
 	}
 
+	cm, err := t.factory.NodeExporterConfigMap()
+	if err != nil {
+		return fmt.Errorf("initializing node-exporter ConfigMap failed: %w", err)
```
Comment (machine424) [inline pkg/tasks/nodeexporter.go:99]:
```suggestion
		return fmt.Errorf("initializing node-exporter accelerators collector ConfigMap failed: %w", err)
```

File: pkg/tasks/nodeexporter.go [hunk 10]
```diff
@@ -94,6 +94,15 @@ func (t *NodeExporterTask) Run(ctx context.Context) error {
...
+	}
+	err = t.client.CreateOrUpdateConfigMap(ctx, cm)
+	if err != nil {
+		return fmt.Errorf("reconciling node-exporter ConfigMap failed: %w", err)
```
Comment (machine424) [inline pkg/tasks/nodeexporter.go:103]:
```suggestion
		return fmt.Errorf("reconciling node-exporter accelerators collector ConfigMap failed: %w", err)
```

File: pkg/tasks/nodeexporter.go [hunk 11]
```diff
@@ -94,6 +94,15 @@ func (t *NodeExporterTask) Run(ctx context.Context) error {
 		return fmt.Errorf("reconciling node-exporter Service failed: %w", err)
 	}
 
+	cm, err := t.factory.NodeExporterConfigMap()
```
Comment (machine424) [inline pkg/tasks/nodeexporter.go:97]:
```suggestion
	cm, err := t.factory.NodeExporterAcceleratorsCollectorConfigMap()
```

**Review by machine424 — CHANGES_REQUESTED** [review]
passing through
left some naming suggestions.

File: jsonnet/versions.yaml (hunk 5 above)
Comment (jan--f) [reply jsonnet/versions.yaml:18]: No need to worry about the `versions` ci job. It just informative for us and not a required check.

Comment (yevgeny-shnaidman) [reply jsonnet/versions.yaml]: Fxied

File: jsonnet/components/node-exporter.libsonnet (hunk 6 above)
Comment (yevgeny-shnaidman) [reply jsonnet/components/node-exporter.libsonnet]: Fixed

File: jsonnet/components/node-exporter.libsonnet (hunk 7 above)
Comment (yevgeny-shnaidman) [reply jsonnet/components/node-exporter.libsonnet]: Fixed

File: pkg/manifests/manifests.go (hunk 8 above)
Comment (yevgeny-shnaidman) [reply pkg/manifests/manifests.go]: Fixed

File: pkg/tasks/nodeexporter.go (hunk 9 above)
Comment (yevgeny-shnaidman) [reply pkg/tasks/nodeexporter.go]: Fixed

File: pkg/tasks/nodeexporter.go (hunk 10 above)
Comment (yevgeny-shnaidman) [reply pkg/tasks/nodeexporter.go]: Fixed

File: pkg/tasks/nodeexporter.go (hunk 11 above)
Comment (yevgeny-shnaidman) [reply pkg/tasks/nodeexporter.go]: Fixed

File: jsonnet/components/node-exporter.libsonnet (hunk 3 above)
Comment (yevgeny-shnaidman) [reply jsonnet/components/node-exporter.libsonnet]: i cannot add the cli arg in the current commit, since it will crush the node-exporter ( the collector.accelerators.configmpa-path flag is not defined there yet. So  in this PR i will just remove the EVN variable, and then add the flag in the node-exporter PR, and then will add another commit in this repo for defining the cli args

File: assets/node-exporter/daemonset.yaml [hunk 12]
```diff
@@ -79,6 +79,9 @@ spec:
         - mountPath: /var/node_exporter/textfile
           name: node-exporter-textfile
           readOnly: true
+        - mountPath: /var/node_exporter/config
```
Comment (machine424) [inline assets/node-exporter/daemonset.yaml:82]:
```suggestion
        - mountPath: /var/node_exporter/accelerators_collector_config
```

File: assets/node-exporter/daemonset.yaml [hunk 13]
```diff
@@ -182,6 +185,12 @@ spec:
       - name: node-exporter-kube-rbac-proxy-config
         secret:
           secretName: node-exporter-kube-rbac-proxy-config
+      - configMap:
+          items:
+          - key: accelerators-collector-config.yaml
+            path: accelerators-collector-config.yaml
```
Comment (machine424) [inline assets/node-exporter/daemonset.yaml:191]:
```suggestion
          - key: config.yaml
            path: config.yaml
```
(given the change, to the folder name's, above)

File: assets/node-exporter/daemonset.yaml [hunk 14]
```diff
@@ -79,6 +79,9 @@ spec:
         - mountPath: /var/node_exporter/textfile
           name: node-exporter-textfile
           readOnly: true
+        - mountPath: /var/node_exporter/config
+          name: node-exporter-config
```
Comment (machine424) [inline assets/node-exporter/daemonset.yaml:83]:
```suggestion
          name: node-exporter-accelerators-collector-config
```

File: assets/node-exporter/daemonset.yaml [hunk 15]
```diff
@@ -182,6 +185,12 @@ spec:
       - name: node-exporter-kube-rbac-proxy-config
         secret:
           secretName: node-exporter-kube-rbac-proxy-config
+      - configMap:
+          items:
+          - key: accelerators-collector-config.yaml
+            path: accelerators-collector-config.yaml
+          name: node-exporter-accelerators-collector-config
+        name: node-exporter-config
```
Comment (machine424) [inline assets/node-exporter/daemonset.yaml:193]:
```suggestion
        name: node-exporter-accelerators-collector-config
```

File: jsonnet/components/node-exporter.libsonnet [hunk 16]
```diff
@@ -350,4 +389,16 @@ function(params)
...
+
+    acceleratorsCollectorConfigmap: {
+      apiVersion: 'v1',
+      kind: 'ConfigMap',
+      metadata: {
+        name: acceleratorsConfigMapName,
+        namespace: cfg.namespace,
+      },
+      data: {
+        'accelerators-collector-config.yaml': std.manifestYamlDoc(acceleratorsConfigData),
```
Comment (machine424) [inline jsonnet/components/node-exporter.libsonnet:401]: we could use `acceleratorsConfigFileName` here I think.

File: jsonnet/components/node-exporter.libsonnet [hunk 17]
```diff
@@ -3,6 +3,26 @@ local textfileVolumeName = 'node-exporter-textfile';
 local tlsVolumeName = 'node-exporter-tls';
 local wtmpPath = '/var/log/wtmp';
 local wtmpVolumeName = 'node-exporter-wtmp';
+local configDir = '/var/node_exporter/config';
+local configVolumeName = 'node-exporter-config';
+local acceleratorsConfigFileName = 'accelerators-collector-config.yaml';
```
Comment (machine424) [inline jsonnet/components/node-exporter.libsonnet:8]:
```suggestion
local acceleratorsConfigFileName = 'config.yaml';
```
see other change suggestions.

File: Documentation/deps-versions.md [hunk 18]
```diff
@@ -1,7 +1,7 @@
...
+| release-4.20 | [0.27.0](https://github.com/openshift/prometheus-alertmanager/blob/release-4.20) | [0.18.2](https://github.com/openshift/kube-rbac-proxy/blob/release-4.20) | [2.13.0](https://github.com/openshift/kube-state-metrics/blob/release-4.20) | [0.7.2](https://github.com/openshift/kubernetes-metrics-server/blob/release-4.20) | [1.0.0](https://github.com/openshift/monitoring-plugin/blob/release-4.20) | [1.8.2](https://github.com/openshift/node_exporter/blob/release-4.20) | [0.11.0](https://github.com/openshift/prom-label-proxy/blob/release-4.20) | [2.55.1](https://github.com/openshift/prometheus/blob/release-4.20) | [0.78.2](https://github.com/openshift/prometheus-operator/blob/release-4.20) | [0.36.1](https://github.com/openshift/thanos/blob/release-4.20) |
```
Comment (machine424) [inline Documentation/deps-versions.md:3]: you could revert this change as well.

**Review by machine424 — COMMENTED** [review]
I had some final naming suggestions. I understand this may seem nitpicky, but I think they'll be helpful to anyone who is not very familiar with node-exporter or the downstream accelerators collector, and who will need to interact with them by providing clear and consistent naming.

This
/lgtm
Thanks for working on this.
/hold
In case the other colleagues have anything to add. 
I understand if you don't want/have time for the suggestions, I can take care of it in another PR.

The failures in okd-scos-e2e-aws-ovn don't seem to be related to the changes. Apparently, the test is flaky (which explains is why it isn't required).

File: jsonnet/components/node-exporter.libsonnet (hunk 17 above)
Comment (yevgeny-shnaidman) [reply jsonnet/components/node-exporter.libsonnet]: fixed

File: jsonnet/components/node-exporter.libsonnet (hunk 16 above)
Comment (yevgeny-shnaidman) [reply jsonnet/components/node-exporter.libsonnet]: in case it is a key, jsonnet does not substitute the variable for its value, but uses the name of the variable directly as key name

Comment (yevgeny-shnaidman) [reply jsonnet/components/node-exporter.libsonnet]: fixed

File: assets/node-exporter/daemonset.yaml (hunk 15 above)
Comment (yevgeny-shnaidman) [reply assets/node-exporter/daemonset.yaml]: fixed

File: assets/node-exporter/daemonset.yaml (hunk 14 above)
Comment (yevgeny-shnaidman) [reply assets/node-exporter/daemonset.yaml]: fixed

File: assets/node-exporter/daemonset.yaml (hunk 13 above)
Comment (yevgeny-shnaidman) [reply assets/node-exporter/daemonset.yaml]: fixed

File: assets/node-exporter/daemonset.yaml (hunk 12 above)
Comment (yevgeny-shnaidman) [reply assets/node-exporter/daemonset.yaml]: fixed

**Comment by machine424** [comment]
/lgtm
thanks again.

**Comment by yevgeny-shnaidman** [comment]
> Thanks, this lgtm overall. I think there are two things we still need (please correct me if I'm missing anything):
> 
> 1. The change for a default config map location in [Add configmap support to the accelerator's collector node_exporter#155](https://github.com/openshift/node_exporter/pull/155)
> 2. And perhaps a test confirming that we remain consistent with the default location? For example create a invalid config map at the default location and expect an error, i.e. fail the test if no error is thrown? For inspiration see https://github.com/openshift/cluster-monitoring-operator/blob/master/test/e2e/node_exporter_test.go.
If i understand correctly , both those issues are relevant for the node exporter PR:
1) default config map location can be added  once the PR node-exporter repo is pushed.
2) The same as above: once we also push the PR for node-exporter , then we can add e2e test

**Comment by machine424** [comment]
As I mentioned earlier, no need to retrigger `okd-scos-e2e-aws-ovn`, it's flaky but it's not blocking the PR as it's not required.

The PR is blocked because of [tide](https://prow.ci.openshift.org/pr?query=is%3Apr+repo%3Aopenshift%2Fcluster-monitoring-operator+author%3Ayevgeny-shnaidman+head%3Ayevgeny%2Fadding-configmap); the PR should be attached to a valid jira ticket https://docs.ci.openshift.org/docs/architecture/jira/#non-bug-jira-references

and the /hold should be removed, cc @jan--f if https://github.com/openshift/cluster-monitoring-operator/pull/2530#issuecomment-2529304530 is ok for you.

**Comment by yevgeny-shnaidman** [comment]
/retitle MGMT-19498: Adding accelerators configuration file to the node-exporter
